bindings/rust/s2n-tls/src/client_hello.rs

```diff
@@ -0,0 +1,401 @@
+    // prevent it from being used externally
+    pub(crate) fn from_ptr(hello: &s2n_client_hello) -> &Self {
+        // safety: see safety justifications [1]
+        unsafe { &*(hello as *const s2n_client_hello as *const ClientHello) }
+    }
+
+    // safety justifications [2]
+    fn deref_mut_ptr(&self) -> *mut s2n_client_hello {
+        self.deref() as *const s2n_client_hello as *mut s2n_client_hello
+    }
+
+    /// internal function which calculates the hash, and also returns the size
+    /// required for the full fingerprint. External customers should instead use
+    /// `fingerprint_hash()` or `fingerprint()`.
+    fn fingerprint_hash_and_size_hint(
+        &self,
+        hash: FingerprintType,
+    ) -> Result<(Vec<u8>, u32), Error> {
+        let mut hash_result = vec![0; MD5_HASH_SIZE as usize];
+        let mut hash_size: u32 = 0;
+        let mut str_size: u32 = 0;
+        unsafe {
+            s2n_client_hello_get_fingerprint_hash(
+                self.deref_mut_ptr(),
+                hash.into(),
+                MD5_HASH_SIZE,
+                hash_result.as_mut_ptr(),
+                &mut hash_size,
+                &mut str_size,
+            )
+            .into_result()?
+        };
+        Ok((hash_result, str_size))
+    }
+
+    /// `fingerprint_string_with_size` will try to compute the fingerprint
+    /// string using a buffer of size `capacity`. This is useful for customers
+    /// who are very concerned with efficiency, since it allows the fingerprint
+    /// string to be calculated in a single pass.
+    pub fn fingerprint_string_with_size(
+        &self,
+        hash: FingerprintType,
+        capacity: u32,
+    ) -> Result<String, Error> {
+        let mut string: Vec<u8> = vec![0; capacity as usize];
+        let mut output_size = 0;
+        unsafe {
+            s2n_tls_sys::s2n_client_hello_get_fingerprint_string(
+                self.deref_mut_ptr(),
+                hash.into(),
+                capacity,
+                string.as_mut_ptr(),
+                &mut output_size,
+            )
+            .into_result()?
+        };
+        return Ok(String::from_utf8(string)
+            .map_err(|parse_error| Error::application(Box::new(parse_error)))?);
+    }
+
+    /// `fingerprint` returns a Vector containing the raw bytes of the hash and
+    /// a String containing the fingerprint string. This function will calculate
+    /// the fingerprint twice. First it will calculate the hash and figure out
+    /// how large the fingerprint string will be, then on the second pass it
+    /// will allocate an appropriately sized buffer and calculate the
+    /// fingerprint string.
+    pub fn fingerprint(&self, hash: FingerprintType) -> Result<(Vec<u8>, String), Error> {
+        let (fingerprint_hash, string_size) = self.fingerprint_hash_and_size_hint(hash)?;
+        let fingerprint_string = self.fingerprint_string_with_size(hash, string_size)?;
```

Comment:
I'm not sure about allocating an arbitrarily large string with no oversight from the application. We might want to enforce a maximum size provided by the application. But at that point, the memory vs computation tradeoffs between this and fingerprint_string_with_size are a little questionable.

Reply:
Is it fair to say arbitrarily large? I know that the comments in `api/unstable/fingerprint.h` called out that there wasn't any specific upper bound on the size of the string, but I'm assuming that the range is distinctly finite?

In my looking it seemed like most fingerprints are less than 512 bytes, which I'd be very comfortable allocating.

I'm assuming that JA3 is well defined enough that there wouldn't be any pathological use cases? E.g. nasty client hello that results in an 18 Gb string. But if there's a way for that to happen then I'd definitely support putting in an upper bound and erroring out if the fingerprint requires more than, e.g. 16 kB of memory.